File: Kattis/metaprogramming.py
import sys


def main():
    d = {}

    for i in sys.stdin:
        q, *x = list(map(str, i.split()))

        if q == 'define':
            d[x[1]] = int(x[0])

        if q == 'eval':
            print('undefined' if not all(k in d for k in (x[0], x[2])) else ('true' if eval(x, d) else 'false'))

def eval(i, d) -> bool:
    a, b, c = i

    return {
        '>': lambda a, b: d[a] > d[b],
        '<': lambda a, b: d[a] < d[b],
        '=': lambda a, b: d[a] == d[b]
    }[b](a, c)

if __name__ == '__main__':
    main()

File: kattis/metaprogramming.rs
use std::{
  collections::HashMap,
  io::{prelude::*, stdin},
};

fn parse(s: String) -> i32 {
  s.trim().parse::<i32>().unwrap()
}

fn main() {
  let mut vars = HashMap::new();

  for line in stdin().lock().lines() {
    let input = line.unwrap();
    let query: Vec<&str> = input.split(" ").collect();

    match query[0] {
      "define" => {
        vars.insert(query[2].to_string(), parse(query[1].to_string()));
      }
      "eval" => {
        if !vars.contains_key(query[1]) || !vars.contains_key(query[3]) {
          println!("undefined");
        } else {
          match query[2] {
            "<" => println!("{}", vars.get(query[1]) < vars.get(query[3])),
            ">" => println!("{}", vars.get(query[1]) > vars.get(query[3])),
            "=" => println!("{}", vars.get(query[1]) == vars.get(query[3])),
            _ => panic!(),
          };
        }
      }
      _ => panic!(),
    }
  }
}
